scipy/sparse/linalg/_interface.py

```diff
@@ -178,8 +178,15 @@ def _init_dtype(self):
         """Called from subclasses at the end of the __init__ routine.
         """
         if self.dtype is None:
-            v = np.zeros(self.shape[-1])
+            v = np.zeros(self.shape[-1], dtype=np.int8)
             self.dtype = np.asarray(self.matvec(v)).dtype
+            # the code below fails
```

Comment:
Delete commented-out code.

Reply:
I was hoping that someone could suggest an easy fix... It would be beneficial to use `matmat` instead of `matvec` and to also involve `rmatmat`

Reply:
Generally, I think we should assume that, when a combination of these methods are provided, that they are consistent with one another. Or people can bypass this and just provide the `dtype` explicitly.

Reply:
Using `matmat` would instantly fix the problem reported in https://github.com/scipy/scipy/issues/19212 ...

If `matmat` is used correctly in the commented out code, but results in errors, there are bugs in the codes giving the errors, see https://github.com/scipy/scipy/issues/19207#issuecomment-1712842372 ?

Reply:
No, it would not have solved the problem in #19212. The problem there was that you had an invalid `LinearOperator`. Even if you got the `dtype` determination to work by avoiding calling `matvec()` *there*, in that one place, that `matvec()` would still be broken elsewhere. The only way to fix #19212 is to fix `_CustomLinearOperator` to accept at least one of `matvec` or `matmat`, as I explained. Then you can implement just `matmat` and get a fully-functional `LinearOperator`.

The errors you report in that comment are exactly the same kind of error: their `matvec` implementations are invalid so calling `matmat()` (from *anywhere*, not just the `dtype` determination) are broken as well.